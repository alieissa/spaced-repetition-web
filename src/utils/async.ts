/** @format */
/**
 * Credit goes to https://medium.com/@fillopeter/pattern-matching-with-typescript-done-right-94049ddd671c
 * For the pattern matching code
 */
export type Untriggered = {
  type: 'Untriggered'
}
export function Untriggered(): Untriggered {
  return {
    type: 'Untriggered',
  }
}

const t: Optional<Answer, 'easiness' | 'quality' | 'interval'> = {
  id: '',
  content: 'fsdfds',
}
export type Loading<V> = {
  type: 'Loading'
  value: V
}
export function Loading<V>(v: V): Loading<V> {
  return {
    type: 'Loading',
    value: v,
  }
}

export type Failure<E> = {
  type: 'Failure'
  value: E
}
export function Failure<E>(e: E): Failure<E> {
  return {
    type: 'Failure',
    value: e,
  }
}

export type Success<V> = {
  type: 'Success'
  value: V
}
export function Success<V>(v: V): Success<V> {
  return {
    type: 'Success',
    value: v,
  }
}

export type Async<L = unknown, F = unknown, S = unknown> =
  | Untriggered
  | Loading<L>
  | Failure<F>
  | Success<S>

export type AsyncStatus<R, E, A> = Async<R, E, A>['type']

export type AsyncMap<R, E, A, U> = {
  [K in AsyncStatus<R, E, A>]: U extends { type: K } ? U : never
}

export type AsyncStatusMap<R, E, A> = AsyncMap<R, E, A, Async<R, E, A>>
export type AsyncPattern<T, R, E, A> = {
  [K in keyof AsyncStatusMap<R, E, A>]: (a: AsyncStatusMap<R, E, A>[K]) => T
}

export function match<R, E, A>(
  e: Async<R, E, A>,
): <T>(pattern: AsyncPattern<T, R, E, A>) => T {
  return <T>(pattern: AsyncPattern<T, R, E, A>) => pattern[e.type](e as any)
}
