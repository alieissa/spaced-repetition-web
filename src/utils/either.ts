/** @format */

/**
 * Credit goes to https://medium.com/@fillopeter/pattern-matching-with-typescript-done-right-94049ddd671c
 * For the pattern matching code
 */
export type Right<V> = Readonly<{
  type: 'Right'
  value: V
}>
export function Right<V>(v: V): Right<V> {
  return {
    type: 'Right',
    value: v,
  }
}

export type Left<E = Error> = Readonly<{
  type: 'Left'
  value: E
}>
export function Left<E>(e: E): Left<E> {
  return {
    type: 'Left',
    value: e,
  }
}

export type Either<L, R> = Left<L> | Right<R>

export type EitherType<L, R> = Either<L, R>['type']
export type EitherMap<L, R, U> = {
  [K in EitherType<L, R>]: U extends { type: K } ? U : never
}
export type EitherTypeMap<L, R> = EitherMap<L, R, Either<L, R>>
export type EitherPattern<T, L, R> = {
  [K in keyof EitherTypeMap<L, R>]: (either: EitherTypeMap<L, R>[K]) => T
}

export function match<L, R>(
  e: Either<L, R>,
): <T>(pattern: EitherPattern<T, L, R>) => T {
  return <T>(pattern: EitherPattern<T, L, R>) => pattern[e.type](e as any)
}
