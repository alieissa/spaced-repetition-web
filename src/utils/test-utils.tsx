/** @format */

import { combineReducers, configureStore } from '@reduxjs/toolkit'
import { render } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import { RouterProvider, createMemoryRouter } from 'react-router-dom'
import decksReducer from 'src/modules/decks/decks.reducer'

// As a basic setup, import your same slice reducers
import { loginReducer } from 'src/modules/login'
import { signupReducer } from 'src/modules/signup'
import { verificationReducer } from 'src/modules/verification'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
// interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
//   preloadedState?: PreloadedState<RootState>
//   store?: AppStore
// }

const reducers = combineReducers({
  login: loginReducer,
  signup: signupReducer,
  verification: verificationReducer,
  decks: decksReducer,
})

const withRedux = (ui: React.ReactElement) => {
  const store = configureStore({ reducer: reducers })
  return <Provider store={store}>{ui}</Provider>
}
const withRouter = (ui: React.ReactElement) => {
  const router = createMemoryRouter([{ element: ui, path: '/' }], {
    initialEntries: ['/'],
    initialIndex: 1,
  })

  return <RouterProvider router={router} />
}
export function renderWithProviders(ui: React.ReactElement) {
  // Return an object with the store and all of RTL's query functions
  return render(withRedux(withRouter(ui)))
}

export const flushPromises = () =>
  new Promise((resolve) => setTimeout(resolve, 0))
