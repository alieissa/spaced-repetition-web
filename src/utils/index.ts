/** @format */

import * as async from './async'
import * as either from './either'

export { async }
export { either }
