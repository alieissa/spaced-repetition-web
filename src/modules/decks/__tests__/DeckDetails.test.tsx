/** @format */

import { faker } from '@faker-js/faker'
import '@testing-library/jest-dom'
import { act, render, screen } from '@testing-library/react'
import { rest } from 'msw'
import { setupServer } from 'msw/node'
import {
  RouteObject,
  RouterProvider,
  createMemoryRouter,
} from 'react-router-dom'

import { flushPromises, withRedux } from 'src/utils/test-utils'
import { DeckDetails, DecksListPage } from '..'

const setupRouter = (routes: RouteObject[], initialEntries: string[]) => {
  const router = createMemoryRouter(routes, { initialEntries })
  return <RouterProvider router={router} />
}
const deckId = faker.string.uuid()
const deckDetailsUrl = `${process.env.REACT_APP_API_ENDPOINT}/decks/${deckId}`
const decksUrl = `${process.env.REACT_APP_API_ENDPOINT}/decks`

// Mounting component this way ensures that the route parameter :deckId
// has id of the deck, i.e. deckId defined above
const mountComponent = () => {
  const initialEntries = ['/decks']
  const routes = [
    { path: '/decks', element: <DecksListPage /> },
    { path: `/decks/:deckId`, element: <DeckDetails /> },
  ]
  return render(withRedux(setupRouter(routes, initialEntries)))
}

const handlers = [
  rest.get(decksUrl, (__, res, ctx) => {
    return res(
      ctx.json([
        {
          id: deckId,
          name: 'Test name5',
          cards: [
            {
              id: 'testUuid',
              question: 'Test question',
              answers: [
                {
                  id: 'testUuid',
                  content: 'Test content',
                },
              ],
            },
          ],
        },
      ]),
      ctx.status(200),
    )
  }),
  rest.get(deckDetailsUrl, (__, res, ctx) => {
    return res(
      ctx.json({
        id: deckId,
        name: 'Test name',
        cards: [
          {
            id: 'testUuid',
            question: 'Test question',
            answers: [
              {
                id: 'testUuid',
                content: 'Test content',
              },
            ],
          },
        ],
      }),
      ctx.status(200),
    )
  }),
]

const server = setupServer(...handlers)
beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

describe('DeckDetails', () => {
  describe('view', () => {
    it('should render correctly', async () => {
      const { asFragment } = mountComponent()
      await act(flushPromises)

      expect(asFragment()).toMatchSnapshot()
    })

    it('should display success', async () => {
      mountComponent()

      await act(flushPromises)

      const testId = 'deck-details-success'
      expect(await screen.findByTestId(testId)).toBeInTheDocument()
    })

    // The DeckDetails component is not rendered independently of DecksListPage
    // so there is no scenario where a call to /decks/:id is made. The moment we
    // decks list then that is used as source of truth for deck details
    // it.skip('should display failure', ...)
  })

  describe('interaction', () => {
    it('should display popup menu when clicking on hamburger menu', async () => {
      // Assemble
      mountComponent()
      await act(flushPromises)

      // Act
      const deckDetailsMenuBtn = await screen.findByTestId(
        'deck-details-dropdown',
      )
      await act(() => deckDetailsMenuBtn.click())

      // Assert
      const deckDetailsMenu = await screen.findByTestId(
        'deck-details-dropdown-menu',
      )
      expect(deckDetailsMenu).toBeVisible()
    })

    it('should open delete confirmation dialog when clicking on delete option', async () => {
      // Assemble
      mountComponent()
      await act(flushPromises)
      const deckDetailsMenuBtn = await screen.findByTestId(
        'deck-details-dropdown',
      )
      await act(() => deckDetailsMenuBtn.click())

      // Act
      const deckMenuDeleteOption = await screen.findByTestId(
        'deck-menu-delete-option',
      )
      await act(() => deckMenuDeleteOption.click())

      // Assert
      const deckDeleteConfirmation = await screen.findByTestId(
        'deck-delete-confirmation-dialog',
      )
      expect(deckDeleteConfirmation).toBeVisible()
    })

    it('should display nothing when decks list is empty', async () => {
      // Assemble
      server.use(
        rest.get(decksUrl, (__, res, ctx) => {
          return res(ctx.json([]), ctx.status(200))
        }),
      )
      mountComponent()

      // Assert
      const deckDetailsComponent = screen.queryByTestId('deck-details')
      expect(deckDetailsComponent).not.toBeInTheDocument()
    })
  })
})
