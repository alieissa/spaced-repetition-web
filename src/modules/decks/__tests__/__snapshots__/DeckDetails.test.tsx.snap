// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeckDetails view should render correctly 1`] = `
<DocumentFragment>
  <div
    data-testid="deck-details-success"
  >
    <div
      actions="[object Object]"
      class="justify-space-between bordered"
      style="padding: 1rem 1rem 1rem 2rem; line-height: 2rem;"
      title="Test name"
    >
      <div
        class="align-center"
      >
        <div
          class="ui header"
          style="font-size: 1.5rem; margin-top: 0px; margin-bottom: 0px;"
        >
          Test name
        </div>
      </div>
      <div>
        <button
          class="ui button"
          data-testid="deck-edit-btn"
          style="border-radius: 0;"
        >
          Edit
        </button>
        <div
          aria-expanded="false"
          class="ui dropdown"
          data-testid="deck-details-dropdown"
          role="listbox"
          tabindex="0"
        >
          <i
            aria-hidden="true"
            class="ellipsis vertical icon"
          />
          <div
            class="menu transition"
            data-testid="deck-details-dropdown-menu"
          >
            <div
              class="item"
              role="option"
            >
              <span
                class="text"
              >
                Test
              </span>
            </div>
            <div
              class="item"
              role="option"
            >
              <span
                class="text"
              >
                Delete
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <main
      class="px-2r"
    >
      <div
        class="flex-column"
        style="margin-top: 1.5rem;"
        title="Cards"
      >
        <div
          class="ui small header"
          style="margin-bottom: 0px; margin-top: 0px;"
        >
          Cards
        </div>
        <div>
          <div
            class="ui celled horizontal list"
            role="list"
          >
            <div
              class="item"
              role="listitem"
              style="padding: 0px;"
            >
              <div
                class="ui card pointer"
                data-testid="deck-details-card-0"
                style="box-shadow: none; border: 1px solid; border-radius: unset;"
              >
                <div
                  class="content"
                >
                  Test question
                </div>
              </div>
            </div>
            <div
              class="item"
              role="listitem"
              style="padding: 0px;"
            >
              <div
                class="ui card pointer"
                style="box-shadow: none; border: 1px solid; border-radius: unset;"
              >
                <div
                  class="content"
                >
                  Add card
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </main>
  </div>
</DocumentFragment>
`;
